File: src/teamlense_modules/agent/services/screenshot.service.ts
import { prisma } from "../../../shared/db/prisma";

type UploadScreenshotPayload = {
  userId: string;
  filePath: string;
  sessionId?: string;
  activeApplication?: string;
  windowTitle?: string;
  domain?: string;
  url?: string;
  employeeName?: string;
  projectName?: string;
  capturedAt: Date;
};

type GetScreenshotsPayload = {
  userId: string;
  sessionId?: string;
  limit: number;
  startDate?: Date;
  endDate?: Date;
};

export const ScreenshotService = {
  async uploadScreenshot(payload: UploadScreenshotPayload) {
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "active_application" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "window_title" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "domain" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "url" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "employee_name" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "project_name" TEXT`);

    const id = crypto.randomUUID();
    const rows = (await prisma.$queryRawUnsafe(
      `INSERT INTO "screenshots"
        ("id", "user_id", "session_id", "file_path", "active_application", "window_title",
         "domain", "url", "employee_name", "project_name", "captured_at", "created_at")
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, NOW())
       RETURNING "id", "user_id" AS "userId", "session_id" AS "sessionId",
                 "active_application" AS "activeApplication", "window_title" AS "windowTitle",
                 "domain", "url", "employee_name" AS "employeeName", "project_name" AS "projectName",
                 "captured_at" AS "capturedAt", "created_at" AS "createdAt"`,
      id,
      payload.userId,
      payload.sessionId ?? null,
      payload.filePath,
      payload.activeApplication ?? null,
      payload.windowTitle ?? null,
      payload.domain ?? null,
      payload.url ?? null,
      payload.employeeName ?? null,
      payload.projectName ?? null,
      payload.capturedAt,
    )) as Array<Record<string, unknown>>;

    return rows[0];
  },

  async getScreenshots(payload: GetScreenshotsPayload) {
    const filters = [
      `"user_id" = $1`,
      payload.sessionId ? `"session_id" = $2` : "",
      payload.startDate ? `"captured_at" >= $${payload.sessionId ? 3 : 2}` : "",
      payload.endDate ? `"captured_at" <= $${payload.sessionId ? (payload.startDate ? 4 : 3) : payload.startDate ? 3 : 2}` : "",
    ].filter(Boolean);

    const values: unknown[] = [payload.userId];
    if (payload.sessionId) values.push(payload.sessionId);
    if (payload.startDate) values.push(payload.startDate);
    if (payload.endDate) values.push(payload.endDate);
    values.push(payload.limit);

    return prisma.$queryRawUnsafe(
      `SELECT "id", "user_id" AS "userId", "session_id" AS "sessionId",
              "active_application" AS "activeApplication", "window_title" AS "windowTitle",
              "domain", "url", "employee_name" AS "employeeName", "project_name" AS "projectName",
              "captured_at" AS "capturedAt", "created_at" AS "createdAt"
       FROM "screenshots"
       WHERE ${filters.join(" AND ")}
       ORDER BY "captured_at" DESC
       LIMIT $${values.length}`,
      ...values,
    );
  },

  async getScreenshotById(id: string) {
    const screenshot = await prisma.screenshot.findUnique({
      where: { id },
    });

    return screenshot;
  },

  async deleteScreenshot(id: string) {
    return prisma.screenshot.delete({
      where: { id },
    });
  },

  async deleteOldScreenshots(daysOld: number = 7) {
    const date = new Date();
    date.setDate(date.getDate() - daysOld);

    return prisma.screenshot.deleteMany({
      where: {
        createdAt: {
          lt: date,
        },
      },
    });
  },
};

File: backend-go/internal/services/screenshot_service.go
package services

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/teamlens/backend-go/internal/models"
)

type ScreenshotService struct {
	pool *pgxpool.Pool
}

func NewScreenshotService(pool *pgxpool.Pool) *ScreenshotService {
	return &ScreenshotService{pool: pool}
}

type UploadScreenshotPayload struct {
	UserID            string
	FilePath          string
	SessionID         *string
	ActiveApplication *string
	WindowTitle       *string
	Domain            *string
	URL               *string
	EmployeeName      *string
	ProjectName       *string
	CapturedAt        time.Time
}

type GetScreenshotsPayload struct {
	UserID    string
	UserIDs   []string // Multiple user IDs (ANY match)
	SessionID *string
	Limit     int
	Offset    int
	StartDate *time.Time
	EndDate   *time.Time
}

func (s *ScreenshotService) UploadScreenshot(ctx context.Context, payload *UploadScreenshotPayload) (*models.Screenshot, error) {
	id := RandomToken(16)

	var s3 struct {
		ID     string
		UserID string
	}

	err := s.pool.QueryRow(ctx,
		`INSERT INTO screenshots
		 (id, user_id, session_id, file_path, active_application, window_title,
		  domain, url, employee_name, project_name, captured_at, created_at)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, NOW())
		 RETURNING id`,
		id, payload.UserID, payload.SessionID, payload.FilePath,
		payload.ActiveApplication, payload.WindowTitle,
		payload.Domain, payload.URL, payload.EmployeeName, payload.ProjectName,
		payload.CapturedAt,
	).Scan(&s3.ID)
	if err != nil {
		return nil, fmt.Errorf("insert screenshot: %w", err)
	}

	return &models.Screenshot{
		ID:                id,
		UserID:            payload.UserID,
		SessionID:         payload.SessionID,
		FilePath:          payload.FilePath,
		ActiveApplication: payload.ActiveApplication,
		WindowTitle:       payload.WindowTitle,
		Domain:            payload.Domain,
		URL:               payload.URL,
		EmployeeName:      payload.EmployeeName,
		ProjectName:       payload.ProjectName,
		CapturedAt:        payload.CapturedAt,
		CreatedAt:         time.Now().UTC(),
	}, nil
}

func (s *ScreenshotService) GetScreenshots(ctx context.Context, payload *GetScreenshotsPayload) ([]models.Screenshot, error) {
	limit := payload.Limit
	if limit <= 0 || limit > 200 {
		limit = 50
	}

	query := `SELECT s.id, s.user_id, s.session_id, s.file_path, s.active_application, s.window_title,
	                 s.domain, s.url, s.employee_name, s.project_name, s.captured_at, s.created_at
	          FROM screenshots s`
	args := []interface{}{}
	paramIdx := 1

	if len(payload.UserIDs) > 0 {
		// Multiple user IDs: WHERE user_id = ANY($1)
		query += fmt.Sprintf(` WHERE s.user_id = ANY($%d)`, paramIdx)
		args = append(args, payload.UserIDs)
		paramIdx++
	} else if payload.UserID != "" {
		query += fmt.Sprintf(` WHERE s.user_id = $%d`, paramIdx)
		args = append(args, payload.UserID)
		paramIdx++
	}

	if payload.SessionID != nil && *payload.SessionID != "" {
		query += fmt.Sprintf(` AND s.session_id = $%d`, paramIdx)
		args = append(args, *payload.SessionID)
		paramIdx++
	}
	if payload.StartDate != nil {
		query += fmt.Sprintf(` AND s.captured_at >= $%d`, paramIdx)
		args = append(args, *payload.StartDate)
		paramIdx++
	}
	if payload.EndDate != nil {
		query += fmt.Sprintf(` AND s.captured_at <= $%d`, paramIdx)
		args = append(args, *payload.EndDate)
		paramIdx++
	}

	query += ` ORDER BY s.captured_at DESC`
	query += fmt.Sprintf(` LIMIT $%d`, paramIdx)
	args = append(args, limit)
	paramIdx++

	if payload.Offset > 0 {
		query += fmt.Sprintf(` OFFSET $%d`, paramIdx)
		args = append(args, payload.Offset)
		paramIdx++
	}

	rows, err := s.pool.Query(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("query screenshots: %w", err)
	}
	defer rows.Close()

	var screenshots []models.Screenshot
	for rows.Next() {
		var ss models.Screenshot
		if err := rows.Scan(&ss.ID, &ss.UserID, &ss.SessionID, &ss.FilePath,
			&ss.ActiveApplication, &ss.WindowTitle, &ss.Domain, &ss.URL,
			&ss.EmployeeName, &ss.ProjectName, &ss.CapturedAt, &ss.CreatedAt); err != nil {
			return nil, fmt.Errorf("scan screenshot: %w", err)
		}
		screenshots = append(screenshots, ss)
	}
	return screenshots, nil
}

func (s *ScreenshotService) GetScreenshotByID(ctx context.Context, id string) (*models.Screenshot, error) {
	var ss models.Screenshot
	err := s.pool.QueryRow(ctx,
		`SELECT id, user_id, session_id, file_path, active_application, window_title,
		        domain, url, employee_name, project_name, captured_at, created_at
		 FROM screenshots WHERE id = $1`, id,
	).Scan(&ss.ID, &ss.UserID, &ss.SessionID, &ss.FilePath,
		&ss.ActiveApplication, &ss.WindowTitle, &ss.Domain, &ss.URL,
		&ss.EmployeeName, &ss.ProjectName, &ss.CapturedAt, &ss.CreatedAt)
	if err != nil {
		return nil, fmt.Errorf("get screenshot: %w", err)
	}
	return &ss, nil
}

func (s *ScreenshotService) DeleteScreenshot(ctx context.Context, id string) error {
	_, err := s.pool.Exec(ctx, `DELETE FROM screenshots WHERE id = $1`, id)
	return err
}

func (s *ScreenshotService) DeleteOldScreenshots(ctx context.Context, daysOld int) error {
	cutoff := time.Now().AddDate(0, 0, -daysOld)
	_, err := s.pool.Exec(ctx, `DELETE FROM screenshots WHERE created_at < $1`, cutoff)
	return err
}

// BuildScreenshotURL constructs the URL path to serve the screenshot file
func BuildScreenshotURL(filePath string) string {
	if strings.HasPrefix(filePath, "/") {
		return "/api/v1/screenshots/serve/" + strings.TrimPrefix(filePath, "/uploads/")
	}
	return "/api/v1/screenshots/serve/" + strings.TrimPrefix(filePath, "uploads/")
}
